### PR TITLE
Implement process_counts in ExpectationMP, VarianceMP and CountsMP

### LINKED ISSUE #5241: Implement `ExpectationMP.process_counts`
# Context

Currently, our sample-based measurement processes that inherit from `SampleMeasurement` implement a [process_samples](https://github.com/PennyLaneAI/pennylane/blob/a05ae286c5925e0c881099cea751193d368779fc/pennylane/measurements/measurements.py#L535) method. This method is capable of turning any samples (array of int64) into the required measurement result. This method is currently used internally by `default.qubit`.

```
@qml.qnode(qml.device('default.qubit'))
def circuit():
    qml.Hadamard(0)
    return qml.sample(wires=(0,1))

samples = circuit(shots=100)
qml.expval(qml.Z(0)).process_samples(samples, wire_order=qml.wires.Wires((0,1)))
```

But external devices and plugins may want to store sampling information in the form of a counts dictionary instead, like `{'00': 45, '01': 55}`. This representation is much more condensed and memory efficient. Because of this, we have recently added a `SampleMeasurement.process_counts` method, and implemented it for `ProbabilityMP`.

# Implementation Details

The task is to implement [ExpectationMP.process_counts(counts, wire_order)](https://github.com/PennyLaneAI/pennylane/blob/b97f7a55c5fcd5b3d0397e9654043b151a7aa0ba/pennylane/measurements/expval.py#L77) in such a way that expectation values can be computed from a counts dictionary.

```pycon
>>> counts = {'000': 100, '100': 100 }
>>> wire_order = qml.wires.Wires((0,1,2))
>>> qml.expval(qml.Z(0)).process_counts(counts, wire_order)
0.0
>>> qml.expval(qml.Z(1)).process_counts(counts, wire_order)
1.0
```

Note that the counts are assumed to be taken in the basis for the provided observable already, with any diagonalizing gates already applied.

# Requirements

A completed PR should:

* Add the `process_counts` method to `ExpecatationMP`
* Add unit tests to [pennylane/tests/measurements/test_expval.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/measurements/test_expval.py)
* Pass black and pylint checks
* Add a changelog entry to [pennylane/releases/changelog-dev.md](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md) under Improvements. Don't forget to add your name to the list of contributors at the bottom :)

### LINKED ISSUE #5244: Implement `VarianceMP.process_counts`
# Context

Currently, our sample-based measurement processes that inherit from `SampleMeasurement` implement a [process_samples](https://github.com/PennyLaneAI/pennylane/blob/a05ae286c5925e0c881099cea751193d368779fc/pennylane/measurements/measurements.py#L535) method. This method is capable of turning any samples (array of int64) into the required measurement result. This method is currently used internally by `default.qubit`.

```
@qml.qnode(qml.device('default.qubit'))
def circuit():
    qml.Hadamard(0)
    return qml.sample(wires=(0,1))

samples = circuit(shots=100)
qml.expval(qml.Z(0)).process_samples(samples, wire_order=qml.wires.Wires((0,1)))
```

But external devices and plugins may want to store sampling information in the form of a counts dictionary instead, like `{'00': 45, '01': 55}`. This representation is much more condensed and memory efficient. Because of this, we have recently added a `SampleMeasurement.process_counts` method, and implemented it for `ProbabilityMP`.

# Implementation Details

The task is to implement [VarianceMP.process_counts(counts, wire_order)](https://github.com/PennyLaneAI/pennylane/blob/e166f7eb69499b043ede63158be1eb95774471a2/pennylane/measurements/var.py#L71) in such a way that expectation values can be computed from a counts dictionary.

```pycon
>>> counts = {'000': 100, '100': 100 }
>>> wire_order = qml.wires.Wires((0,1,2))
>>> qml.var(qml.Z(0)).process_counts(counts, wire_order)
1.0
>>> qml.var(qml.Z(1)).process_counts(counts, wire_order)
0.0
```

Note that the counts are assumed to be taken in the basis for the provided observable already, with any diagonalizing gates already applied.

# Requirements

A completed PR should:

* Add the `process_counts` method to `VarianceMP`
* Add unit tests to [pennylane/tests/measurements/test_varpy](https://github.com/PennyLaneAI/pennylane/blob/master/tests/measurements/test_var.py)
* Pass black and pylint checks
* Add a changelog entry to [pennylane/releases/changelog-dev.md](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md) under Improvements. Don't forget to add your name to the list of contributors at the bottom :)

### LINKED ISSUE #5249: Implement `CountsMP.process_counts`
# Context

Currently, our sample-based measurement processes that inherit from `SampleMeasurement` implement a [process_samples](https://github.com/PennyLaneAI/pennylane/blob/a05ae286c5925e0c881099cea751193d368779fc/pennylane/measurements/measurements.py#L535) method. This method is capable of turning any samples (array of int64) into the required measurement result. This method is currently used internally by `default.qubit`.

```
@qml.qnode(qml.device('default.qubit'))
def circuit():
    qml.Hadamard(0)
    return qml.sample(wires=(0,1))

samples = circuit(shots=100)
qml.counts(wires=0, all_outcomes=True).process_samples(samples, wire_order=qml.wires.Wires((0,1)))
```

But external devices and plugins may want to store sampling information in the form of a counts dictionary instead, like `{'00': 45, '01': 55}`. This representation is much more condensed and memory efficient. Because of this, we have recently added a `SampleMeasurement.process_counts` method, and implemented it for `ProbabilityMP`.

# Implementation Details

The task is to implement [CountsMP.process_counts(counts, wire_order)](https://github.com/PennyLaneAI/pennylane/blob/3407a9e3fd76f0b112ef22e14ef8a4ba3d5a339f/pennylane/measurements/counts.py#L159) in such a way that a counts dictionary can be processed in such a way that agrees with the specification for the measurement process.

This includes:

* Adding in all outcomes if `all_outcomes=True`
* Removing entries with a value of `0` if `all_outcomes=False`
* Tracing over wires

```pycon
>>> counts = {'000': 100, '100': 100 }
>>> wire_order = qml.wires.Wires((0,1,2))
>>> qml.counts(wires=(0,1), all_outcomes=True).process_counts(counts, wire_order)
{'00': 100, '01': 0, '10': 100, '11': 0}
>>> qml.counts(wires=(0,2), all_outcomes=False).process_counts(counts, wire_order)
{'00': 100, '10': 10}
>>> counts2 = {'00': 0, '10': 0, '01': 0, '11': 50}
>>> qml.counts(wires=0, all_outcomes=False).process_counts(counts2, wire_order=qml.wires.Wires((0,1)))
{'1': 50}
```

Note that the counts are assumed to be taken in the basis for the provided observable already, with any diagonalizing gates already applied.

**Warning:** This processing should leave the input dictionary unchanged.

# Requirements

A completed PR should:

* Add the `process_counts` method to `CountsMP`
* Add unit tests to [pennylane/tests/measurements/test_counts.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/measurements/test_counts.py)
* Pass black and pylint checks
* Add a changelog entry to [pennylane/releases/changelog-dev.md](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md) under Improvements. Don't forget to add your name to the list of contributors at the bottom :)

### PR DESCRIPTION
**Pull request by Tarun-Kumar07**
**Context:**
Under #4941 abstract method `process_counts` was added to `SampleMeasurement`. This method provides support to  process counts dictionary produced by external systems.

**Description of the Change:**
* Implement `process_counts` in `ExpectationMP`, `VarianceMP` and `CountsMP`.
* While implementing `process_counts` in `CountsMP` some common logic was extracted from `ProbabilityMP` to `CountsMP`

**Benefits:**
The below methods won't throw `NotImplementedError` exception
* `ExpectationMP.process_counts`
* `VarianceMP.process_counts`
* `CountsMP.process_counts`

**Possible Drawbacks:**
* All classes which inherit from `SampleMeasurement` implement `process_counts` except `SampleMP`. 
Should I add an implementation for that ?
* After implementing `process_counts` in all child classes can we make the method abstract similar to `process_samples` ? This might break some tests where some classes inherit from `SampleMeasurement` as there `process_counts` is not implemented
* It is assumed that `counts` dictionary is valid and caller is responsible to call with valid dictionary. It has already been discussed in this [conversation](https://github.com/PennyLaneAI/pennylane/pull/4952/files/eb9c1ee81ea87c274a5ec094f90a0065d05fdc36#r1433117236) and this is a design choice.

**Related GitHub Issues:**
This PR
* fixes #5249 
* fixes #5244 
* fixes #5241

***

**Further details**
I'm excited to be making my first open-source contribution with this PR 😄 . 
As a newcomer to the community, I'm eager to learn and improve. 
Any feedback for enhancements would be greatly appreciated!

**Internal Shortcut Stories**
- [sc-57166]
- [sc-57274]
- [sc-57307]

### CODE REVIEW

**Comment by trbromley** [comment]
Thanks a lot for opening this @Tarun-Kumar07! We'll give this a review soon and also help with the CI issue you mentioned. Sorry if things are a little delayed this week, we're busy preparing for the PennyLane 0.35 release next week :rocket:

**Comment by Tarun-Kumar07** [comment]
@trbromley Thank you for your response.
If there's anything I can do to help with the PennyLane 0.35 release preparations, please let me know. 
I'd be more than happy to assist in any way I can! 🚀

**Comment by timmysilv** [comment]
hi @Tarun-Kumar07 , thanks for bearing with us! I've pushed a fix for the strange "Documentation Check" issue to the master branch. If you update your fork with the latest master, it should disappear 🤞 give it a try, I'll keep an eye out to make sure it's resolved here.

File: pennylane/measurements/probs.py
Comment (albi3ro) [inline pennylane/measurements/probs.py:247]:
```suggestion
       with qml.QueuingManager.stop_recording():
           helper_counts = qml.counts(wires=self.wires, all_outcomes=False)
       mapped_counts = helper_counts.process_counts(counts, wire_order)
```

File: pennylane/measurements/var.py
Comment (albi3ro) [inline pennylane/measurements/var.py:133]:
```suggestion
        with qml.QueuingManager.stop_recording():
            prob = qml.probs(wires=self.wires).process_counts(counts=counts, wire_order=wire_order)
```

**Review by albi3ro — CHANGES_REQUESTED** [review]
Sorry about the delay in review.

This is looking really great @Tarun-Kumar07 :)

I only have two main requests:

1) stripping out zeros entries if `all_outcomes=False`
2) Using `qml.QueuingManager.stop_recording()` around any location that creates an intermediary measurement process

But once those are done, I'll be happy to approve.

**Comment by Tarun-Kumar07** [comment]
Hey @albi3ro,

I noticed that all classes which inherit from SampleMeasurement implement process_counts except SampleMP. Should I add an implementation for that?

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:117]:
```suggestion
* Implemented the method `process_counts` in `ExpectationMP`, `VarianceMP`, and `CountsMP`.
  [(#5256)](https://github.com/PennyLaneAI/pennylane/pull/5256)
```

We link the pull request, rather than the issue.

**Review by albi3ro — APPROVED** [review]
Just left a request to change the changelog entry to reference this PR, rather than the issues.

Looking great to me 🎉

Yes, we could also use `SampleMP.process_counts`, but we could potentially stick that into an independent PR instead.

File: doc/releases/changelog-dev.md
Comment (trbromley) [inline doc/releases/changelog-dev.md:160]: Note that you can put your name in alphabetical order, according to last name.

**Comment by Tarun-Kumar07** [comment]
Hi @albi3ro and @trbromley,

Thanks for reviewing my PR. 
It's my first open-source contribution, and I'm eager to contribute even more 🎉.

Appreciate it!

**Comment by Tarun-Kumar07** [comment]
@albi3ro, regarding adding `SampleMP.process_counts`. 
Should I create a PR directly or make an issue about it? 
What's the preferred process?

**Review by astralcai — APPROVED** [review]
Looks Good! 🎸

**Comment by albi3ro** [comment]
> @albi3ro, regarding adding `SampleMP.process_counts`. Should I create a PR directly or make an issue about it? What's the preferred process?

I'll go ahead and create an issue right.

Since this PR now has two approvals 🎉 , I'll go ahead and enable auto-merge.  Congrats on the PR!

**Pull request merged**
